### PR TITLE
#72 Handle error 429 and Retry-After header

### PR DESCRIPTION
**Pull request by parsley72**
Add error 429 to isRetryableError().
Add retryAfter() to handle Retry-After header.
Handle negative delay from retryDelay() as cancellation.
Add example for checking Retry-After against limit.

### CODE REVIEW

File: es/index.js [hunk 1]
```diff
@@ -140,6 +157,23 @@ function fixConfig(axios, config) {
  *   return retryCount * 1000;
  * }});
  *
+ * // Custom retry which checks Retry-After header against limit
+ * axiosRetry(axios, { retryDelay : (retryCount, error) => {
+ *   if (error.response) {
+ *     const retry_after = error.response.headers["retry-after"];
+ *     if (retry_after) {
+ *       // check if retry is less than 5 seconds
```
Comment (parsley72) [inline es/index.js:165]: Ideally I'd like to compare to `axios.timeout` but I couldn't figure out how to do that.

**Comment by Uzlopak** [comment]
This is imho a wrong implementation.

axios.timeout is not relevant for Error 429. Error 429 means, that the server has too many requests and in retry-after-header you get two types of responses. A datetime object or an delay in seconds

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After

So a potential solution would be actually (it is typescript, so just remove it to have just javascript)

```typescript
export async function handleRetryAfter(error: AxiosError, resolve: <T = any>(value?: T | PromiseLike<T>) => void): Promise<void> {
	if (error.response) {
		const retryAfter = error.response.headers["retry-after"];
		let retryAfterInMs = 0;

		if (/\d+/.test(retryAfter)) {
			retryAfterInMs = retryAfter * 1000;
		}

		if (
			typeof retryAfter === "string" &&
			isNaN(new Date(retryAfter as string).getTime()) === false
		) {
			retryAfterInMs = new Date(retryAfter).getTime() - (error.config as AxiosRetryRequestConfig).lastRequestTime;
		}
		if (retryAfterInMs) {
			await new Promise(resolve => setTimeout(resolve, retryAfterInMs));
		}
	}
	resolve();
}
```

and in the response interceptor you should then change to 

```typescript
			return new Promise(
				resolve =>
					new Promise(resolveRetryAfter => handleRetryAfter(error, resolveRetryAfter)
						.then(() => setTimeout(() => resolve(axios(config)), delay))));
```

So you resolveRetryAfter after the corresponding retry and then you do the actual call.

**Comment by parsley72** [comment]
What I needed was an error 429 handler which would use Retry-After but only to a point - the particular API I'm calling can return a Retry-After up to 15 mins long. I don't want my page to wait that long, I want to have a 5 second cutoff, so I need a mechanism to cancel the retry. Without that I don't need to change anything in axios-retry, you can just use https://github.com/softonic/axios-retry/issues/72#issuecomment-699785860.

You're right about handling the date case, I hadn't noticed that (since the API I'm using only returns the seconds case).

**Comment by Uzlopak** [comment]
Then you have to actually not set a timeout but introduce another option e.g. maxRetryAfter, which handles the corresponding abort condition, when the retry would take too long.

**Comment by parsley72** [comment]
Why? I don't think you're reading the PR properly - the new `retryAfter()` function just checks for a Retry-After value. The implementation that includes a timing limit is just a comment that's provided as an example - you can change the value to whatever you want.

**Comment by Uzlopak** [comment]
Yes you are right. I implemented it now differently. Thank you for your clarification

**Comment by parsley72** [comment]
Updated with Date handling, minor fixes.

**Comment by parsley72** [comment]
@Uzlopak thoughts?

**Comment by Uzlopak** [comment]
Well like I wrote I implemented it differently.

I used your code as base and did it like this:


```
/**
 * The Retry-After response HTTP header indicates how long the user agent 
 * should wait before making a follow-up request. There are three main cases 
 * this header is used:
 *
 * - When sent with a 503 (Service Unavailable) response, this indicates how
 *   long the service is expected to be unavailable.
 * - When sent with a 429 (Too Many Requests) response, this indicates how 
 *   long to wait before making a new request.
 * - When sent with a redirect response, such as 301 (Moved Permanently), this
 *   indicates the minimum time that the user agent is asked to wait before 
 *   issuing the redirected request.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
 */
export function retryAfterDelay(response: { headers?: { "retry-after": string } }, lastRequestTime: number): number {
	let retryAfterInMs = 0;

	if (response) {
		const retryAfter = response.headers["retry-after"];

		if (/\d+/.test(retryAfter)) {
			retryAfterInMs = ~~retryAfter * 1000;
		}

		if (
			typeof retryAfter === "string" &&
			isNaN(new Date(retryAfter as string).getTime()) === false
		) {
			retryAfterInMs = new Date(retryAfter).getTime() - lastRequestTime;
		}
	}

	return retryAfterInMs;
}

```

and in our axiosRetry adaptation I use it like this: 

```

				if (shouldRetry) {
					request.retryCount += 1;
					const delay = retryAfterDelay(error.response, request.lastRequestTime) || retryDelay(request.retryCount, error);
```

So I dont wait additionally to the retryDelay but instead of the retryDelay. This makes more sense, as it would not add up the delays created by the retryDelay method.

**Pull request closed without merging**